libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,319 @@
 package bridge
 
 import (
-	"bytes"
 	"context"
 	"errors"
 	"fmt"
 	"net"
 
 	"github.com/containerd/log"
+	"github.com/docker/docker/libnetwork/iptables"
 	"github.com/docker/docker/libnetwork/netutils"
+	"github.com/docker/docker/libnetwork/portallocator"
+	"github.com/docker/docker/libnetwork/portmapper"
 	"github.com/docker/docker/libnetwork/types"
-	"github.com/ishidawataru/sctp"
 )
 
-func (n *bridgeNetwork) allocatePorts(ep *bridgeEndpoint, reqDefBindIP net.IP, ulPxyEnabled bool) ([]types.PortBinding, error) {
-	if ep.extConnConfig == nil || ep.extConnConfig.PortBindings == nil {
-		return nil, nil
-	}
+type portBinding struct {
+	*types.PortBinding
+	proxy portmapper.UserlandProxy
+}
 
-	defHostIP := net.IPv4zero // 0.0.0.0
-	if reqDefBindIP != nil {
-		defHostIP = reqDefBindIP
+// allocatePorts takes cfg, the configuration for port mappings, selects host
+// ports when ranges are given, starts docker-proxy or its dummy to reserve
+// host ports, and sets up iptables NAT/forwarding rules as necessary. If
+// anything goes wrong, it will undo any work it's done and return an error.
+// Otherwise, the returned slice of portBinding has an entry per address
+// family (if cfg describes a mapping for 'any' host address, it's expanded
+// into mappings for IPv4 and IPv6, because that's how the mapping is presented
+// in 'inspect'). HostPort and HostPortEnd in each returned portBinding are set
+// to the selected and reserved port.
+func (n *bridgeNetwork) allocatePorts(
```

Comment:
It's the original name, and it's doing exactly what it did before. It's good that that's more obvious now.

It can be renamed. Maybe `addPortMappings`?

Reply:
Renamed.